docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,551 @@
++		if containsResourceRef {
++			condition.SetReferenceResolved(desiredCopy, corev1.ConditionFalse,
++				&condition.ReferenceNotResolvedMessage,
++				&condition.ReferenceNotResolvedReason)
++		}
++		return desiredCopy, err
++	}
+
++	// If the desired resource had references, set the ReferenceResolved
++	//condition to true
++	if containsResourceRef {
++		condition.SetReferenceResolved(dereferencedDesired, corev1.ConditionTrue,
++			&condition.ReferenceResolvedMessage,
++			&condition.ReferenceResolvedReason)
++	} else {
++		condition.RemoveReferenceResolved(dereferencedDesired)
++	}
++	desired = dereferencedDesired
+
+	rlog.Enter("rm.ReadOne")
+	latest, err = rm.ReadOne(ctx, desired)
+	rlog.Exit("rm.ReadOne", err)
+	if err != nil {
+		if err != ackerr.NotFound {
+			return latest, err
+		}
+		if isAdopted {
+			return nil, ackerr.AdoptedResourceNotFound
+		}
+		if latest, err = r.createResource(ctx, rm, desired); err != nil {
+			return latest, err
+		}
+	} else {
+		if latest, err = r.updateResource(ctx, rm, desired, latest); err != nil {
+			return latest, err
+		}
+	}
+	// Attempt to late initialize the resource. If there are no fields to
+	// late initialize, this operation will be a no-op.
+	if latest, err = r.lateInitializeResource(ctx, rm, latest); err != nil {
+		return latest, err
+	}
+	return r.handleRequeues(ctx, latest)
+}
+```
+
+#### Service Controller Generated Code
+**generator.yaml**
+
+NOTE: The VpcLink configuration below is just for generating sample code for
+list for references. Using APIID as SecurityGroupId is not a legitimate example.
+This is due to the restriction that in first revision we are only allowing referencing
+the resource from same api-group.
+```yaml
+resources:
+  Integration:
+    fields:
+      Api:
+        references:
+          kind: API
+          field_path: Status.APIID
+          target_field: APIID
+          is_list: false
+          required: true
+      ApiId:
+        is_required: false
+  VpcLink:
+    fields:
+      SecurityGroups:
+        references:
+          kind: API
+          field_path: Status.APIID
+          target_field: SecurityGroupIDs
+          is_list: true
+          required: false
+```
+
+**integration/manager.go**
+```go
+package integration
+
+import (
+...
+
++	acksvcv1alpha1 "github.com/aws-controllers-k8s/apigatewayv2-controller/apis/v1alpha1"
+	ackv1alpha1 "github.com/aws-controllers-k8s/runtime/apis/core/v1alpha1"
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+...
+	svcsdk "github.com/aws/aws-sdk-go/service/apigatewayv2"
+	svcsdkapi "github.com/aws/aws-sdk-go/service/apigatewayv2/apigatewayv2iface"
++	"k8s.io/apimachinery/pkg/types"
++	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// +kubebuilder:rbac:groups=apigatewayv2.services.k8s.aws,resources=integrations,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apigatewayv2.services.k8s.aws,resources=integrations/status,verbs=get;update;patch
```

Comment:
this would be required once we support cross-controller/cross-apigroup references, but this is not necessary given that a controller automatically has read permissions for all resources it manages.